plugins/amqp_compressed_proof_plugin/amqp_compressed_proof_plugin.cpp

```diff
@@ -0,0 +1,393 @@
+            cb.second(cc.generate_serialized_proof(action_return_value_active_this_block));
+      }
+   }
+
+   my->reversible_action_entries_index.get<generator_impl::by_block_num>().erase(
+           my->reversible_action_entries_index.get<generator_impl::by_block_num>().begin(),
+           my->reversible_action_entries_index.get<generator_impl::by_block_num>().upper_bound(bsp->block_num)
+   );
+}
+
+void compressed_proof_generator::add_result_callback(action_filter_func&& filter, merkle_proof_result_func&& result) {
+   my->callbacks.emplace_back(std::make_pair(filter, result));
+}
+
+struct amqp_compressed_proof_plugin_impl {
+   std::unique_ptr<compressed_proof_generator> proof_generator;
+   std::list<reliable_amqp_publisher>          publishers;
+};
+
+amqp_compressed_proof_plugin::amqp_compressed_proof_plugin():my(new amqp_compressed_proof_plugin_impl()){}
+amqp_compressed_proof_plugin::~amqp_compressed_proof_plugin() = default;
+
+void amqp_compressed_proof_plugin::set_program_options(options_description& cli, options_description& cfg) {
+   cfg.add_options()
+         ("compressed-proof", bpo::value<std::vector<std::string>>()->composing(),
+          "Define a filter, AMQP server, and exchange to publish compressed proofs to. May be specified multiple times. "
+          "Must be in the form of:\n"
+          "  <name>=<filter>=<server>=<exchange>\n"
+          "Where:\n"
+          "   <name>     \tis a unique name identifying this proof; used for persistent filenames\n"
+          "   <filter>   \tis a comma separated list of account names to filter actions on\n"
+          "   <server>   \tis the AMQP server URI\n"
+          "   <exchange> \tis the AMQP exchange to publish to\n"
+          )
+         ;
+
+   cli.add_options()
+         ("compressed-proof-delete", bpo::bool_switch(),
+          "Delete all compressed proof data files for unconfirmed messages and blocks");
+}
+
+void amqp_compressed_proof_plugin::plugin_initialize(const variables_map& options) {
+   try {
+      boost::filesystem::path dir = app().data_dir() / "amqp";
+      boost::system::error_code ec;
+      boost::filesystem::create_directories(dir, ec);
+
+      const std::string file_prefix = "cproof";
+
+      chain::controller& controller = app().get_plugin<chain_plugin>().chain();
+
+      if(options.count("compressed-proof-delete") && options.at("compressed-proof-delete").as<bool>())
+         for(const auto& p : boost::filesystem::directory_iterator(dir))
+            if(boost::starts_with(p.path().filename().generic_string(), file_prefix))
+               boost::filesystem::remove(p.path(), ec);
+
+      const boost::filesystem::path reversible_blocks_file = dir / (file_prefix + "-reversible.bin"s);
+      my->proof_generator = std::make_unique<compressed_proof_generator>(reversible_blocks_file);
+      controller.irreversible_block.connect([&generator=*my->proof_generator,&controller](const chain::block_state_ptr& bsp) {
```

Comment:
Doesn't this need to keep a connection, so we can disconnect on shutdown?

Reply:
By the time shutdown() is called the event loop is stopped so controller will never pump out more blocks (or other signals) so the connection to the now destroyed amqp_compressed_proof_plugin will never be called.